### PR TITLE
Photo naming conventions

### LINKED ISSUE #69: Redesign photo downloading API to have a custom naming parameter
Ideally users will be able to specify a naming scheme as the parameter (e.g. passing `"MHM_{date}_{direction}..etc"` as a parameter) with the default set to the current setup.

### PR DESCRIPTION
**Pull request by pingle14**
Fixes #69 

- added None-arg handling in photo_download
- added photo naming customization
- added photo_download tests
- added command-line flags to customize photo names
- Updated pdocs to include guide for photo naming flags

### CODE REVIEW

**Comment by pingle14** [comment]
`test_country_download(protocol, prefix)` is failing due to `df.drop(["SHAPE"], axis=1, inplace=True)` in `geoenrich.py`.

SHAPE is not found in axis=1. -- Could this be a change in `arcgis.features.GeoAccessor`, where the returned dataframe no longer contains a SHAPE field?

**Comment by pingle14** [comment]
I added an if-statement checking the existance of `SHAPE` before dropping the column. Now the same test is failing due to issues with `larvae_to_num` in `mhm.py`

Specifically, a `ValueError: cannot call 'vectorize' on size 0 inputs unless 'otypes' is set` is thrown. 

See screenshot for details
larvae_conversion_issue

File: setup.py [hunk 1]
```diff
@@ -57,7 +57,7 @@
     long_description_content_type="text/markdown",
     author="IGES",
     maintainer="Matteo Kimura, Prachi Ingle, Pratham Babaria",
-    maintainer_email="mateus.sakata@gmail.com, phna14@gmail.com, prathambabaria1@gmail.com",
+    maintainer_email="mateus.sakata@gmail.com, prachi.ingle@utexas.edu, prathambabaria1@gmail.com",
```
Comment (Piphi5) [inline setup.py:60]: Can you do this with another PR (and also maybe take out Pratham since he's not working on this), it's not really related to the photo stuff?

File: go_utils/tests/photo_download_test.py [hunk 2]
```diff
@@ -68,48 +70,168 @@
...
 @pytest.mark.parametrize(
-    "input_file, func, desired",
+    "input_file, func, excluded_fields, additional_info, desired",
     [
         (
             "go_utils/tests/sample_data/mhm_small.csv",
             get_mhm_download_targets,
-            desired_mhm_names,
+            [],
+            "",
```
Comment (Piphi5) [inline go_utils/tests/photo_download_test.py:117]: Why are you passing blank lists and strings?

File: go_utils/photo_download.py [hunk 3]
```diff
@@ -354,6 +462,13 @@ def download_lc_photos(
         The column name of the column that contains the east photo urls. If not specified, these photos will not be included.
     west_photo : str, default = "lc_WestPhotoUrl"
         The column name of the column that contains the west photo urls. If not specified, these photos will not be included.
+    exclude_from_name : list of str, default=[]
+        A list of coloumn names to exclude from the downloaded photo names.
+        Accepted Excluded Names include: "direction", "latitude", "longitude", "date_str", "lc_id"
+    additional_name_stem : str, default=""
+        Additional custom information the user can add to the name.
+        NOTE: Name format is as follows: "lc_{additional_name_stem}_{direction}_{latitude}_{longitude}_{date_str}_{lc_id}_{photo_id}.png"
+            Excluded name fields will not appear in the name
```
Comment (Piphi5) [inline go_utils/photo_download.py:471]: Could we (this would kind of be a major change) instead have like a format string set up so that the user could cherry-pick exactly which fields they'd like: e.g. `(f"{latitude}_{longitude}_{direction}", latitude="lc_Latitude")` and then have default arguments for those things?

File: go_utils/photo_download.py [hunk 4]
```diff
@@ -48,11 +54,22 @@ def download_photo(url, directory, filename):
     filename : str
         The name of the photo
     """
-    downloaded_obj = requests.get(url, allow_redirects=True)
-    filename = remove_bad_characters(filename)
-    out_path = os.path.join(directory, filename)
-    with open(out_path, "wb") as file:
-        file.write(downloaded_obj.content)
+    if any(x is not None for x in [url, directory, filename]):
```
Comment (Piphi5) [inline go_utils/photo_download.py:57]: I think for nonetype checking, it might be more robust to use `pd.isna()` (applies to other usages if None)
```suggestion
    if any(not pd.isna(x) for x in [url, directory, filename]):
```

File: go_utils/_cli.py [hunk 5]
```diff
@@ -212,23 +258,53 @@ def lc_photo_download():
...
+    name_group.add_argument(
+        "--nargs_exclude",
+        "-x",
+        nargs="+",
+        help="Exclude these fields from photo names.",
+        choices=lc_name_fields,
+        type=str,
+        action="store",
+    )
+
```
Comment (Piphi5) [inline go_utils/_cli.py:307]: Again, see comment for photo url format string naming

**Review by Piphi5 — CHANGES_REQUESTED** [review]

File: setup.py (hunk 1 above)
Comment (pingle14) [reply setup.py:60]: done!

File: go_utils/tests/photo_download_test.py (hunk 2 above)
Comment (pingle14) [reply go_utils/tests/photo_download_test.py:117]: oo I wanted other tests to call this test function `test_all_field_naming`, so I passed blank lists (indicating no field should be excluded from the name) and "" (indicating no additional info was passed to the name). This allows me to test both extremes: including all fields, no fields, and some fields.

This test may slightly change if I restructure the API to ask users to specify what to Include, rather than exclude.

File: go_utils/photo_download.py (hunk 3 above)
Comment (pingle14) [reply go_utils/photo_download.py]: done! -- I changed the logic to instead ask for fields to INCLUDE in the name, in any order that the user would like. I also updated the flags + tests + documentation to reflect this.

However, I'm not sure allowing the flexibility of `latitude='lc_latitude'` is the best approach, since the column name was already specified in the `get_lc_download_targets` method. I think it's safe to assume the same column will contain latitude information.

File: go_utils/photo_download.py (hunk 4 above)
Comment (pingle14) [reply go_utils/photo_download.py]: done! -- Thanks for the tip!

File: go_utils/_cli.py (hunk 5 above)
Comment (pingle14) [reply go_utils/_cli.py]: updated!

**Comment by pingle14** [comment]
`isort` and `black` seem to have conflicting standards for code formatting -- specifically when tabbing multiple imports

File: go_utils/photo_download.py [hunk 6]
```diff
@@ -220,6 +288,11 @@ def download_mhm_photos(
         The column name of the column that contains the watersource photo urls. If not specified, the larvae photos will not be included.
     abdomen_photo : str, default = "mhm_AbdomenCloseupPhotoUrls"
         The column name of the column that contains the abdomen photo urls. If not specified, the larvae photos will not be included.
+    include_in_name : list of str, default=[]
+        A list of coloumn names to include into the downloaded photo names. The order of items in this list is maintained in the outputted name list of coloumn names to include into the downloaded photo names
```
Comment (Piphi5) [inline go_utils/photo_download.py:292]: Typo for `column names`

File: go_utils/photo_download.py [hunk 7]
```diff
@@ -286,15 +366,39 @@ def get_lc_download_targets(
...
-        date_str = pd.to_datetime(str(date)).strftime("%Y-%m-%d")
-        photo_id = get_globe_photo_id(url)
+        if not pd.isna(url) and "https" in url:
+            date_str = pd.to_datetime(str(date)).strftime("%Y-%m-%d")
+            photo_id = get_globe_photo_id(url)
 
-        name = f"lc_{direction}_{round(latitude, 5)}_{round(longitude, 5)}_{date_str}_{lc_id}_{photo_id}.png"
-        name = remove_bad_characters(name)
-        targets.add((url, directory, name))
+            # f"lc_{direction}_{round(latitude, 5)}_{round(longitude, 5)}_{date_str}_{lc_id}_{photo_id}.png"
```
Comment (Piphi5) [inline go_utils/photo_download.py:373]: You can take out this comment

File: go_utils/tests/photo_download_test.py [hunk 8]
```diff
@@ -69,48 +71,168 @@
...
+            [],
+            "ADDITIONAL",
+            [(re.sub(r"mhm\_.*\_", "mhm_ADDITIONAL_", x)) for x in full_mhm_names],
+        ),
+        (
+            "go_utils/tests/sample_data/lc_small.csv",
+            get_lc_download_targets,
+            [],
+            "ADDITIONAL",
+            [(re.sub(r"lc\_.*\_", "lc_ADDITIONAL_", x)) for x in full_lc_names],
```
Comment (Piphi5) [inline go_utils/tests/photo_download_test.py:195]: If additional info just adds a string that doesn't necessarily relate with each individual entry, is it worth adding?

File: go_utils/photo_download.py [hunk 9]
```diff
@@ -286,15 +366,39 @@ def get_lc_download_targets(
...
+                "lc_id": lc_id,
+            }
+
+            if not pd.isna(photo_id) and int(photo_id) >= 0:
+                name = build_name()
+                targets.add((url, directory, name))
+            else:
+                print("Invalid photoID: " + photo_id)
+        else:
+            print("Invalid URL: " + str(url))
```
Comment (Piphi5) [inline go_utils/photo_download.py:401]: See warning comment.

File: go_utils/photo_download.py [hunk 10]
```diff
@@ -325,6 +429,8 @@ def download_lc_photos(
     south_photo="lc_SouthPhotoUrl",
     east_photo="lc_EastPhotoUrl",
     west_photo="lc_WestPhotoUrl",
+    include_in_name=[],
+    additional_name_stem="",
```
Comment (Piphi5) [inline go_utils/photo_download.py:433]: See additional name comment in the tests

File: go_utils/photo_download.py [hunk 11]
```diff
@@ -276,6 +351,11 @@ def get_lc_download_targets(
         The column name of the column that contains the east photo urls. If not specified, these photos will not be included.
     west_photo : str, default = "lc_WestPhotoUrl"
         The column name of the column that contains the west photo urls. If not specified, these photos will not be included.
+    include_in_name : list of str, default=[]
+        A list of coloumn names to include into the downloaded photo names. The order of items in this list is maintained in the outputted name
+        Accepted Included Names include: "direction", "latitude", "longitude", "date_str", "lc_id"
+    additional_name_stem : str, default=""
+        Additional custom information the user can add to the name.
```
Comment (Piphi5) [inline go_utils/photo_download.py:358]: See additional name comment in the tests

**Review by Piphi5 — CHANGES_REQUESTED** [review]
A few minor changes and a few questions. But I think we're close to finishing up so great work!

File: go_utils/photo_download.py (hunk 3 above)
Comment (Piphi5) [reply go_utils/photo_download.py]: That's true, I think this current setup works!

File: go_utils/tests/photo_download_test.py (hunk 8 above)
Comment (pingle14) [reply go_utils/tests/photo_download_test.py:195]: I think so -- For example, perhaps a User wants to download 2 mutually exclusive sets of images. The additional Information parameter could help the User organize those 2 sets.

File: go_utils/photo_download.py (hunk 9 above)
Comment (pingle14) [reply go_utils/photo_download.py:401]: see https://github.com/IGES-Geospatial/globe-observer-utils/pull/73#discussion_r776763168

File: go_utils/photo_download.py (hunk 10 above)
Comment (pingle14) [reply go_utils/photo_download.py:433]: see https://github.com/IGES-Geospatial/globe-observer-utils/pull/73#discussion_r776763168

File: go_utils/photo_download.py (hunk 11 above)
Comment (pingle14) [reply go_utils/photo_download.py:358]: see https://github.com/IGES-Geospatial/globe-observer-utils/pull/73#discussion_r776763168

File: go_utils/info.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+
+__doc__ = """
+Useful stored values for GLOBE Observer scripts
+
+Stored Values
+-------------
+- start_date: the default starting date of GLOBE Obsever data requests in YYYY-MM-DD Form.
+- end_date: the current date in YYYY-MM-DD Form.
+- regions_dict: contains all the different GLOBE regions and the countries associated with each region
+"""
```
[the remark this answers is not in the document]
Comment (pingle14) [reply go_utils/info.py]: done!

File: go_utils/photo_download.py (hunk 7 above)
Comment (pingle14) [reply go_utils/photo_download.py]: done!

File: go_utils/photo_download.py [hunk 13]
```diff
@@ -120,6 +150,11 @@ def get_mhm_download_targets(
         The column name of the column that contains the watersource photo urls. If not specified, the larvae photos will not be included.
     abdomen_photo : str, default = "mhm_AbdomenCloseupPhotoUrls"
         The column name of the column that contains the abdomen photo urls. If not specified, the larvae photos will not be included.
+    include_in_name : list of str, default=[]
+        A list of coloumn names to include into the downloaded photo names. The order of items in this list is maintained in the outputted name
```
[the remark this answers is not in the document]
Comment (pingle14) [reply go_utils/photo_download.py]: thanks!

File: go_utils/photo_download.py [hunk 14]
```diff
@@ -145,19 +180,50 @@ def get_photo_args(
...
-                continue
-            photo_id = get_globe_photo_id(url)
 
-            classification = genus
-            if pd.isna(classification):
-                classification = "None"
-            elif not pd.isna(species):
-                classification = f"{classification} {species}"
-            name = f"mhm_{url_type}_{watersource}_{round(latitude, 5)}_{round(longitude, 5)}_{date_str}_{mhm_id}_{classification}_{photo_id}.png"
+        # f"mhm_{url_type}_{watersource}_{round(latitude, 5)}_{round(longitude, 5)}_{date_str}_{mhm_id}_{classification}_{photo_id}.png"
```
[the remark this answers is not in the document]
Comment (pingle14) [reply go_utils/photo_download.py]: done!

File: go_utils/photo_download.py [hunk 15]
```diff
@@ -49,11 +57,22 @@ def download_photo(url, directory, filename):
...
+    else:
+        print(
+            "Either url ("
+            + url
+            + "), directory ("
+            + directory
+            + "), or filename ("
+            + filename
+            + ") was None."
+        )
```
[the remark this answers is not in the document]
Comment (pingle14) [reply go_utils/photo_download.py]: sounds good! -- Thanks for the tip!

File: go_utils/photo_download.py [hunk 16]
```diff
@@ -65,19 +84,28 @@ def download_all_photos(targets):
     targets : list of tuple of str
         Contains tuples that store the url, directory, and filename of the desired photos to be downloaded in that order.
     """
-    for target in targets:
-        download_photo(*target)
+    expectedNumParams = 3
+    if pd.isna(targets):
+        print("ERR: Targets was none")
```
[the remark this answers is not in the document]
Comment (pingle14) [reply go_utils/photo_download.py]: done!

File: go_utils/photo_download.py (hunk 6 above)
Comment (pingle14) [reply go_utils/photo_download.py]: fixed

File: go_utils/photo_download.py (hunk 9 above)
Comment (pingle14) [reply go_utils/photo_download.py]: for this warning -- rather than warning for each invalid photo, I provided a warning at the end of the iteration reporting the number of invalid/skipped photos -- This way there is less console output for the User

File: go_utils/tests/photo_download_test.py (hunk 8 above)
Comment (Piphi5) [reply go_utils/tests/photo_download_test.py:195]: Ah that makes sense

File: go_utils/__init__.py [hunk 17]
```diff
@@ -94,12 +94,18 @@
...
 For example, `mhm_photo_download "input csv path" "output directory"` will take the Mosquito Habitat Mapper data stored in the input CSV and download the photos to the output directory. If the output directory doesn't exist, it will be created.
 
 #### All Flag
 You can use the `--all` or `-a` to set all the following flags.
 
+#### Custom Photo Naming Flags
+* Use `--name_additional` or `-add` to add custom additional information to the photo names.
+* Use `--nargs_include` or `-in` to include a field from the default photo name. Can include any of the following:
+   * For **mosquito_habitat_mapper** -- Accepted Included Names include: `url_type`, `watersource`, `latitude`, `longitude`, `date_str`, `mhm_id`, `classification`
+   * For **land_cover** -- Accepted Included Names include: `direction`, `latitude`, `longitude`, `date_str`, `lc_id`
```
Comment (Piphi5) [inline go_utils/__init__.py:107]: It might be nice to explain what each of these mean (even if some are self explanatory from the name).

File: go_utils/photo_download.py [hunk 18]
```diff
@@ -49,11 +58,15 @@ def download_photo(url, directory, filename):
     filename : str
         The name of the photo
     """
-    downloaded_obj = requests.get(url, allow_redirects=True)
-    filename = remove_bad_characters(filename)
-    out_path = os.path.join(directory, filename)
-    with open(out_path, "wb") as file:
-        file.write(downloaded_obj.content)
+    if any(not pd.isna(x) for x in [url, directory, filename]):
```
Comment (Piphi5) [inline go_utils/photo_download.py:61]: This is really clever

File: go_utils/photo_download.py [hunk 19]
```diff
@@ -276,6 +348,11 @@ def get_lc_download_targets(
         The column name of the column that contains the east photo urls. If not specified, these photos will not be included.
     west_photo : str, default = "lc_WestPhotoUrl"
         The column name of the column that contains the west photo urls. If not specified, these photos will not be included.
+    include_in_name : list of str, default=[]
+        A list of column names to include into the downloaded photo names. The order of items in this list is maintained in the outputted name
+        Accepted Included Names include: "direction", "latitude", "longitude", "date_str", "lc_id"
```
Comment (Piphi5) [inline go_utils/photo_download.py:353]: Add short descriptions in parenthesis for each of these

File: go_utils/photo_download.py [hunk 20]
```diff
@@ -355,6 +460,11 @@ def download_lc_photos(
...
```
Comment (Piphi5) [inline go_utils/photo_download.py:465]: Add brief descriptions

File: go_utils/photo_download.py [hunk 21]
```diff
@@ -220,6 +285,11 @@ def download_mhm_photos(
         The column name of the column that contains the watersource photo urls. If not specified, the larvae photos will not be included.
     abdomen_photo : str, default = "mhm_AbdomenCloseupPhotoUrls"
         The column name of the column that contains the abdomen photo urls. If not specified, the larvae photos will not be included.
+    include_in_name : list of str, default=[]
+        A list of column names to include into the downloaded photo names. The order of items in this list is maintained in the outputted name list of column names to include into the downloaded photo names
+        Accepted Included Names include: "url_type", "watersource", "latitude", "longitude", "date_str", "mhm_id", "classification"
```
Comment (Piphi5) [inline go_utils/photo_download.py:290]: Short description

File: go_utils/photo_download.py [hunk 22]
```diff
@@ -120,6 +144,11 @@ def get_mhm_download_targets(
         The column name of the column that contains the watersource photo urls. If not specified, the larvae photos will not be included.
     abdomen_photo : str, default = "mhm_AbdomenCloseupPhotoUrls"
         The column name of the column that contains the abdomen photo urls. If not specified, the larvae photos will not be included.
+    include_in_name : list of str, default=[]
+        A list of column names to include into the downloaded photo names. The order of items in this list is maintained in the outputted name
+        Accepted Included Names include: "url_type", "watersource", "latitude", "longitude", "date_str", "mhm_id", "classification"
```
Comment (Piphi5) [inline go_utils/photo_download.py:149]: Short description

File: go_utils/photo_download.py [hunk 23]
```diff
@@ -65,19 +78,28 @@ def download_all_photos(targets):
...
+        return None
     return (
         "".join(s.capitalize() + " " for s in name.split("_"))
         .replace("Photo", "")
         .strip()
     )
 
 
-def get_mhm_download_targets(
+def get_mhm_download_targets(  # noqa: C901
```
Comment (Piphi5) [inline go_utils/photo_download.py:102]: You could get around C901 by making "private functions" (see _format_param_name()) outside of this method. It might also read better if you do it that way.

File: go_utils/photo_download.py [hunk 24]
```diff
@@ -233,7 +303,7 @@ def download_mhm_photos(
     return targets
 
 
-def get_lc_download_targets(
+def get_lc_download_targets(  # noqa: C901
```
Comment (Piphi5) [inline go_utils/photo_download.py:306]: See previous comment about C901

**Review by Piphi5 — CHANGES_REQUESTED** [review]

File: go_utils/photo_download.py [hunk 25]
```diff
@@ -284,17 +361,42 @@ def get_lc_download_targets(
...
+                "lc_id": lc_id,
+            }
+
+            if not pd.isna(photo_id) and int(photo_id) >= 0:
+                name = build_name()
+                targets.add((url, directory, name))
+            else:
+                num_invalid_photos + 1
+        else:
+            num_invalid_photos + 1
```
Comment (Piphi5) [inline go_utils/photo_download.py:399]: Not sure if this increments num_invalid_photos. It might also be good to add [tests ](https://docs.pytest.org/en/6.2.x/warnings.html#warns) to make sure this works
```suggestion
                num_invalid_photos += 1
        else:
            num_invalid_photos += 1
```

File: go_utils/photo_download.py [hunk 26]
```diff
@@ -145,19 +175,50 @@ def get_photo_args(
...
+                        "mhm_id": mhm_id,
+                        "classification": classification,
+                    }
+
+                    name = build_name()
+                    targets.add((url, directory, name))
+                else:
+                    num_invalid_photos + 1
+            else:
+                num_invalid_photos + 1
```
Comment (Piphi5) [inline go_utils/photo_download.py:221]: See previous comment about this

**Review by Piphi5 — CHANGES_REQUESTED** [review]

File: go_utils/photo_download.py (hunk 21 above)
Comment (pingle14) [reply go_utils/photo_download.py]: done!

File: go_utils/photo_download.py (hunk 23 above)
Comment (pingle14) [reply go_utils/photo_download.py:102]: updated! -- Thanks for the tip!

File: go_utils/photo_download.py (hunk 25 above)
Comment (pingle14) [reply go_utils/photo_download.py]: added tests! -- Python doesn't allow modifications to an outer variable from within a nested function, so I found a workaround by creating a 1-element dict of the `num_invalid_photos`

File: go_utils/photo_download.py [hunk 27]
```diff
@@ -380,27 +397,13 @@ def get_lc_download_targets(  # noqa: C901
     """
     arguments = locals()
     targets = set()
-    num_invalid_photos = 0
+    num_invalid_photos = {"count": 0}
```
Comment (Piphi5) [inline go_utils/photo_download.py:400]: What's the advantage of keeping it in a dict?

**Review by Piphi5 — CHANGES_REQUESTED** [review]
Just one more comment to be addressed

File: go_utils/photo_download.py (hunk 27 above)
Comment (pingle14) [reply go_utils/photo_download.py:400]: Python doesn't allow modifications to an outer variable from within a nested function, so I found a workaround by creating a 1-element dict of the num_invalid_photos -- Here's a [source](https://stackoverflow.com/questions/11987358/why-nested-functions-can-access-variables-from-outer-functions-but-are-not-allo).

Comment (Piphi5) [reply go_utils/photo_download.py:400]: Would it be possible to return two things (targets, num invalid) from that function and increment the variable that way?

Comment (pingle14) [reply go_utils/photo_download.py]: Sure -- we could return `num_invalid_photos` from the nested function and then sum the vectorized output from `get_mhm/lc_photo_args`.

Alternatively, with the dictionary, we can provide the User with more details on the types of photos skipped (see new commit) with messages like `Skipped 5 invalid photos: {'invalid_URL': 3, 'rejected': 1, 'pending': 1, 'bad_photo_id': 0}`. This could help the user better understand their dataset.

Comment (Piphi5) [reply go_utils/photo_download.py]: I think the latter would be really nice. So we can keep the dictionary as long as we do it like that

Comment (pingle14) [reply go_utils/photo_download.py]: Cool beans! -- It's implemented in the most recent commit :)

**Review by Piphi5 — APPROVED** [review]
